common/transport/amqp/src/amqp.ts

```diff
@@ -132,6 +88,185 @@ export class Amqp {
+                return null;
+              })
+              .catch((err) => {
+                this._amqp.removeListener(_amqpClientError, connectErrorHandler);
+                this._fsm.transition('disconnected');
+                /*Codes_SRS_NODE_COMMON_AMQP_16_003: [The `connect` method shall call the `done` callback if the connection fails.] */
+                this._safeCallback(connectCallback, connectError || err);
+              });
+          },
+          disconnect: (callback) => callback(null, new results.Disconnected()),
+          attachSenderLink: (endpoint, linkOptions, callback) => {
+            this._fsm.handle('connect', (err) => {
+              if (err) {
+                callback(err);
+              } else {
+                this._fsm.handle('attachSenderLink', endpoint, linkOptions, callback);
+              }
+            });
+          },
+          attachReceiverLink: (endpoint, linkOptions, callback) => {
+            this._fsm.handle('connect', (err) => {
+              if (err) {
+                callback(err);
+              } else {
+                this._fsm.handle('attachReceiverLink', endpoint, linkOptions, callback);
+              }
+            });
+          },
+          detachSenderLink: (endpoint, callback) => this._safeCallback(callback),
+          detachReceiverLink: (endpoint, callback) => this._safeCallback(callback),
+          initializeCBS: (callback) => {
+            this._fsm.handle('connect', (err) => {
+              if (err) {
+                callback(err);
+              } else {
+                this._fsm.handle('initializeCBS', callback);
+              }
+            });
+          },
+          putToken: (audience, token, callback) => {
+            this._fsm.handle('initializeCBS', (err) => {
+              if (err) {
+                callback(err);
+              } else {
+                this._fsm.handle('putToken', audience, token, callback);
+              }
+            });
+          },
+          '*': () => this._fsm.deferUntilTransition('connected')
+        },
+        connecting: {
+          connect: () => this._fsm.deferUntilTransition(),
+          disconnect: () => this._fsm.deferUntilTransition(),
+          '*': () => this._fsm.deferUntilTransition('connected')
+        },
+        connected: {
+          connect: (callback) => callback(null, new results.Connected()),
+          disconnect: (disconnectCallback) => {
+            this._fsm.transition('disconnecting');
+            this._disconnect((err) => {
+              this._fsm.transition('disconnected');
+              disconnectCallback(err);
+            });
+          },
+          initializeCBS: (callback) => {
+            this._cbs = new ClaimsBasedSecurityAgent(this._amqp);
+            this._cbs.attach(callback);
+          },
+          putToken: (audience, token, callback) => {
+            this._cbs.putToken(audience, token, callback);
+          },
```

Comment:
the problem is that authentication is managed at the higher layer (device/service AMQP class). mainly because the "authenticated" state varies from one to the other (one has x509 support but not the other for example). 
I think of this layer as mapping to the AMQP primitives. I wish I could displace the send to the link object and thenonly have APIs to create/destroy links and high level objects (CBS, etc) here and then let the higher layer deal with authentication completely but that'd be a breaking change.

it's clearly not perfect and I've gone back and forth on this quite a bit - i may still change stuff as I refactor the higher layer.

---
In reply to: [128129311](https://github.com/Azure/azure-iot-sdk-node/pull/55#discussion_r128129311) [](ancestors = 128129311)